python/lsst/meas/algorithms/astrometrySourceSelector.py

```diff
@@ -0,0 +1,121 @@
+        - sourceCat  a catalog of sources
+        """
+        self._getSchemaKeys(sourceCat.schema)
+
+        result = table.SourceCatalog(sourceCat.table)
+        for source in sourceCat:
+            if self._isGood(source) and not self._isBad(source):
+                result.append(source)
+        return Struct(sourceCat=result)
+
+    def _getSchemaKeys(self, schema):
+        """Extract and save the necessary keys from schema with asKey."""
+        self.parentKey = schema["parent"].asKey()
+        self.centroidKey = table.Point2DKey(schema["slot_Centroid"])
+        self.centroidFlagKey = schema["slot_Centroid_flag"].asKey()
+        self.nChildKey = schema["deblend_nChild"].asKey()
+
+        self.edgeKey = schema["base_PixelFlags_flag_edge"].asKey()
+        self.interpolatedCenterKey = schema["base_PixelFlags_flag_interpolatedCenter"].asKey()
+        self.saturatedKey = schema["base_PixelFlags_flag_saturated"].asKey()
+
+        fluxPrefix = "slot_%sFlux_" % (self.config.sourceFluxType,)
+        self.fluxField = fluxPrefix + "flux"
+        self.fluxKey = schema[fluxPrefix + "flux"].asKey()
+        self.fluxFlagKey = schema[fluxPrefix + "flag"].asKey()
+        self.fluxSigmaKey = schema[fluxPrefix + "fluxSigma"].asKey()
+
+    def _isMultiple(self, source):
+        """Return True if source is likely multiple sources."""
+        if (source.get(self.parentKey) != 0) or (source.get(self.nChildKey) != 0):
+            return True
+        footprint = source.getFootprint()
+        return footprint is not None and len(footprint.getPeaks()) > 1
+
+    def _hasCentroid(self, source):
+        """Return True if the source has a valid centroid"""
+        centroid = source.get(self.centroidKey)
+        return np.all(np.isfinite(centroid)) and not source.getCentroidFlag()
+
+    def _goodSN(self, source):
+        """Return True if source has Signal/Noise > config.minSnr."""
+        return (self.config.minSnr <= 0 or
+                (source.get(self.fluxKey)/source.get(self.fluxSigmaKey) > self.config.minSnr))
+
+    def _isUsable(self, source):
+        """
+        Return True if the source is usable for matching, even if it may have a poor centroid.
+
+        For a source to be usable it must:
+        - have a valid centroid
+        - not be deblended
+        - have a valid flux (of the type specified in this object's constructor)
+        - have adequate signal-to-noise
+        """
+        return self._hasCentroid(source) \
+            and not self._isMultiple(source) \
+            and not source.get(self.fluxFlagKey) \
+            and self._goodSN(source)
+
+    def _isGood(self, source):
+        """
+        Return True if source is usable for matching and likely has a good centroid.
+
+        The additional tests for a good centroid, beyond isUsable, are:
+        - not interpolated in the center (this includes saturated sources,
+            so we don't test separately for that)
```

Comment:
Comment was in the original code I stole this from (measOptimisticB). I added the separate saturated check, since I'm using fake data. I think I'd rather leave it in, to catch corner cases (plus, I don't know anything about how interpolatedCenter and saturated interact beyond that one comment).